pandas/core/generic.py

```diff
@@ -8988,26 +8992,40 @@ def _align_series(
                     other.index, how=join, level=level, return_indexers=True
                 )
 
-            left = self._reindex_indexer(join_index, lidx, copy)
+            if is_series:
+                left = self._reindex_indexer(join_index, lidx, copy)
+            elif join_index is None:
+                left = self.copy() if copy else self
+            else:
+                data = {
+                    c: algos.take_nd(
+                        self[c]._values,
```

Comment:
this will cause a problem if columns are not unique. is that case ruled out here?

Reply:
It is not. Haven't thought of that. I have added a test for that case and also implemented a solution it. However, I am not sure if the BlockManager method, that you are referring to is better.

Reply:
Made some further simplifications and added some more tests